### ISSUE TITLE
Document list of zone IDs

### ISSUE THREAD

**Issue by DiskJunky** [nodatime/nodatime.org#10]
One of the key things listed in the example sitting front and center on the main page of the NodaTime website, is the presence of a hardcoded string for the timezone provider ID. Finding a list of these possible values available online was surprisingly difficult to track down.

There is a GitHub post by user jrolstad here [https://gist.github.com/jrolstad/5ca7d78dbfe182d7c1be](url) which provides the code to generate the list and the list's output.

However, given that this is such a key thing to use the NodaTime library, the Zone ID list should be very accessible and findable on the NodaTime website.

**Comment by jskeet** [comment]
There's a prototype of this up at http://nodatime.org/IanaTimeZones now. This isn't linked anywhere yet, and the URL may well change - I've put it up to get feedback, basically. That primarily means you @DiskJunky, as the original requester :)

**Comment by DiskJunky** [comment]
**Things to close the ticket;**
* None :-) requested data is present and correct

**"Nice to have"s, if they're quick wins;**
* A small note to what Lat/Long measurement (e.g., decimal/radians), is used. I'm assuming decimal, greenwich is 0.
* Simple search capability on data in the table (e.g., like JS datatable)
* Lat/Long should be separated out into different columns (scraping/importing the data becomes easier)
* A note that using lat/long to determine zone can be problematic (https://stackoverflow.com/a/16086964/1838819)
* The UTC offset for each zone (and if possible, if DST applies)


**Stretch items (probably tickets in their own right);**
* The NodaTime version that supports the IANA version (possible in brackets in the drop down)
* Some simple examples of correctly determining the zone ID based on country code, region/state, etc. Probably worth a small article in and of itself.

**Comment by jskeet** [comment]
Nice to hear that the tzvalidate link is useful - that was the last thing I added :) At some point I could potentially add a UI for that, but I expect to be the main user myself, and I'm happy to just put query parameters on :)

Easy things:

- Note about lat/long (possibly at the bottom, with a link at the top to get to the notes)
- Note about using lat/long being problematic to use

The other nice to haves:

- The search part would probably have to be a separate contribution, as I know nothing about it. I'm not a web developer, as should be obvious from the state of that page ;)
- Rather than separating out the lat/long into separate columns, would it be useful to have a "download as csv" link? Or as JSON, perhaps? It's easy to generate the right information, and I'd have thought either of those formats would be easier than scraping.
- UTC offset is tricky, as it's not as simple as "if DST applies"... we could give "the UTC offset right now" or even "the set of UTC offsets that apply in a particular interval, e.g. 2010-2040" if that would be useful. But countries start and stop using DST, so it's not really a binary thing. And Ireland's DST isn't what most people would expect (as it observes DST of -1 hour in winter; summer is its standard time). Let me know what you think on that one.

The stretch items:

- All versions of Noda Time *support* all versions of time zone data - the file format has grown since 1.0, but it's still backward-compatible. (Would a link to download the nzd file for a given version be useful? It's trivial to add.) If you were interested in "which version of Noda Time *shipped* with which version of IANA" that's doable but I'm not sure how useful it is. (The reverse mapping is already on the version history page.)
- The bit about determining the zone ID does sound like a separate page, either a blog post or in the main documentation - we could link to it from the notes though.

I'm definitely not going to close this issue yet, as we'll want a bit more fine tuning and then prominent linking before it's really useful to everyone. Please keep giving feedback - this is really handy :)

**Comment by DiskJunky** [comment]
Fair enough, though I did find the tzvalidate data useful to gain context around the complexity of the zones and their management (plus, I'm nosy :) ).

Easy things;
* The search is definitely in the "nice to have but not necessary" category as the data is searchable via the browser find - it was more of a usability thing. I wouldn't worry about it.
* You're right, having a JSON object returned would be more consumable if anyone was scraping the data. I had a look through the IANA info in my initial search and those downloads were not easily disseminated or consumed (or at least I found it so).
* Fair points on UTC/DST, my suggestion was somewhat ill thought through there. My thinking on DST was more along the lines of a flag to indicate "DST applies to this timezone over the course of a year", rather than an up to date indicate of what is currently in effect for that timezone. Doing the latter would perhaps be a bit of a project in and of itself.

Stretch items;
* Ah, hadn't twigged the backward compatibility feature. As for which version shipped, it is relevant in terms of system maintenance. E.g., in systems we're currently using, these will have to be updated with the latest version of NodaTime to support the EU changes coming in in the next two years. Knowing which version _didn't_ support a timezone change would be fairly critical in assessing systems affected, though I appreciate it wouldn't be a common/frequent ask.

Happy to give feedback, I'll do my best to keep it to the point :)

**Comment by jskeet** [comment]
This is all good feedback, thanks.

- I'll get to work on the JSON side. I suspect that will be pretty trivial. Definitely useful info to have.
- I can see the "set of offsets used over the course of some interval" being useful, and if there's more than one listed then it's *likely* that it's due to DST. Interested users can always click through to the tzvalidate info :)

For the maintenance aspect, if you change your code so that you can supply it with a new nzd file whenever you want to, that *should* work for all versions of Noda Time, and gives you independence between Noda Time version and time zone data version - which I think is a very valuable thing to have anyway.

**Comment by jskeet** [comment]
Okay, sorry for the delay - I've now implemented everything we talked about, I think. Could you have another look and let me know if there's anything else (including any desirable JSON changes) before I start linking to it? I'm not sure about the current URL (https://nodatime.org/IanaTimeZones) - any thoughts on that would be welcome too. (Just TimeZones perhaps?)

**Comment by DiskJunky** [comment]
Nothing to be sorry about - you're at this even over a weekend!

The changes look great, general feedback;
* Some of the notes might be worth adding to the top of the page as well as adding an anchor link to the rest at the bottom as they are currently. I had a feedback point typed out before I realised the info was at the bottom of the page. Definite usability tweak warranted there (though, yes, I should have viewed the _entire_ page before writing feedback);

> I'm guessing UTC? I'm left with no idea of the significance of "2000-2039", even after taking another look at the tzvalidate data. I think all that's needed here is a note of explanation or a rewording - it's presence is fine.


* The column name "Offsets (years 2000-2039)" had me unclear on what it's displaying until I got to the bottom. Possible tweak to "UTC Offset". I'm not sure if the 2000-2039 is worth putting in as it only confuses things - especially if it's an arbitrary cut off. The notes can cover it. 
* The JSON data looks great, very consumable. Coupled with the more human readable tzvalidate link, it's a great resource to have to back up the page. On your point of my work asks on maintenance, it's a nice resource to potentially use to update NodaTime in the manner suggested. Could be a nice NodaTime plugin if requirements to update proves popular generally as this goes wider.
* On the URL, I'd suggest "TimeZones". Most people wouldn't have heard of IANA unless they delved into this whole domain. You know programmers, they usually just want stuff to work :) There are plenty of other links and references throughout the site if they're truly interested.

**Comment by jskeet** [comment]
Hopefully the final set of replies :)

- I've added links to the notes in the table columns for lat/long and offsets. I'd rather avoid putting too many notes at the top, as after you know what you're looking at, they're a distraction for every further visit. It's a definite balancing act between "first time" and "later" users.
- I've renamed "Offsets (years 2000-2039)" to "Modern offsets" with the notes link. Let me know if the notes themselves need further expansion - and I'm still willing to be convinced about the need to have the notes at the top, but I'm *hoping* the link is good enough.
- I'm not sure what you mean about the maintenance side of things. The normal approach to maintaining the time zone database would either be to make sure you're using the latest version (currently only 2.4.x is being updated when there's a new IANA release, but that's only because it's the only one with a significant number of users who use updates; when 3.0 launches, I expect to keep updating 2.4.x for example) or to write code to check https://nodatime.org/tzdb/latest.txt - when the URL returned by that changes, you download that new file and start using that. If things are still unclear, it may be worth asking a Stack Overflow question or raising a new issue.
- URL is now TimeZones. It's linked in various places now, but it's easy to add more links if you'd like...
- "More than JSON" support - I hadn't particularly planned on including even the JSON :) I'm happy to add more formats if there's demand, but I probably won't do so preemptively.
- I'm using .NET Core 3.0.100-preview3-010431 (as per global.json). Perhaps you have an older 3.0 preview installed?
